core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -132,9 +139,22 @@ class ReplicaFetcherThread(name: String,
 
     brokerTopicStats.updateReplicationBytesIn(records.sizeInBytes)
 
+    log.maybeUpdateHighWatermark(partitionData.highWatermark).foreach { newHighWatermark =>
```

Comment:
Note that we are changing the order here as `updateHighWatermark` was called before `maybeIncrementLogStartOffset`. It may be better to keep the original order in order to avoid unexpected side effects. It probably does not matter but I rather prefer to be safe.